include/envoy/network/transport_socket.h

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+#include "envoy/network/connection.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * Action that should occur on a connection after I/O.
+ */
+enum class PostIoAction {
+  // Close the connection.
+  Close,
+  // Keep the connection open.
+  KeepOpen
+};
+
+/**
+ * Result of each I/O event.
+ */
+struct IoResult {
+  PostIoAction action_;
+
+  /**
+   * Number of bytes processed by the I/O event.
+   */
+  uint64_t bytes_processed_;
+};
+
+/**
+ * Callbacks used by transport socket instances to communicate with connection.
+ */
+class TransportSocketCallbacks {
+public:
+  virtual ~TransportSocketCallbacks() {}
+
+  /**
+   * @return int the file descriptor associated with the connection.
+   */
+  virtual int fd() PURE;
+
+  /**
+   * @return Network::Connection& the connection interface.
+   */
+  virtual Network::Connection& connection() PURE;
+
+  /**
+   * @return bool whether the read buffer should be drained. This is used to enforce yielding for
+   *         configured read limits.
+   */
+  virtual bool shouldDrainReadBuffer() PURE;
+
+  /**
+   * Mark read buffer ready to read in the event loop. This is used when yielding following
+   * shouldDrainReadBuffer().
+   */
+  virtual void setReadBufferReady() PURE;
+
+  /**
+   * Raise a connection event to the connection. This can be used by a secure socket (e.g. TLS)
+   * to raise a connected event when handshake is done.
+   * @param event supplies the connection event
+   */
+  virtual void raiseEvent(ConnectionEvent event) PURE;
+};
+
+/**
+ * A transport socket that does actual read / write. It can also do some transformations on
+ * the data (e.g. TLS).
+ */
+class TransportSocket {
+public:
+  virtual ~TransportSocket() {}
+
+  /**
+   * Called by connection once to initialize the transport socket callbacks that the transport
+   * socket should use.
+   * @param callbacks supplies the callbacks instance.
+   */
+  virtual void setTransportSocketCallbacks(TransportSocketCallbacks& callbacks) PURE;
+
+  /**
+   * @return std::string the protocol to use as selected by network level negotiation. (E.g., ALPN).
+   *         If network level negotation is not supported by the connection or no protocol
+   *         has been negotiated the empty string is returned.
+   */
+  virtual std::string protocol() const PURE;
+
+  /**
+   * @return bool whether the socket can be flush closed.
```

Comment:
"can be flushed and closed"  since I'm not sure flush closed is a thing :-)

I'm kind of bummed we can't infer this from Connected.  It'd work for TLS since we always raiseEvent Connected() right after the handshake is complete, but it might cause problems for TCP, as the socket isn't flagged connected until it processes the initial event.   Matt, think this is worth fixing?

Reply:
@lizan the question is for @mattklein123 but can you still update the comment in the interim?

Reply:
Done